homeassistant/helpers/unit_system.py

```diff
@@ -0,0 +1,97 @@
+"""Unit system helper class and methods."""
+
+import logging
+from homeassistant.const import (
+    TEMP_CELSIUS, TEMP_FAHRENHEIT, LENGTH_CENTIMETERS, LENGTH_METERS,
+    LENGTH_KILOMETERS, LENGTH_INCHES, LENGTH_FEET, LENGTH_YARD, LENGTH_MILES,
+    VOLUME_LITERS, VOLUME_MILLILITERS, VOLUME_GALLONS, VOLUME_FLUID_OUNCE,
+    MASS_GRAMS, MASS_KILOGRAMS, MASS_OUNCES, MASS_POUNDS, SYSTEM_METRIC,
+    SYSTEM_IMPERIAL)
+
+_LOGGER = logging.getLogger(__name__)
+
+TYPE_LENGTH = 'length'  # type: str
+TYPE_MASS = 'mass'  # type: str
+TYPE_VOLUME = 'volume'  # type: str
+TYPE_TEMPERATURE = 'temperature'  # type: str
+
+LENGTH_UNITS = [
+    LENGTH_MILES,
+    LENGTH_YARD,
+    LENGTH_FEET,
+    LENGTH_INCHES,
+    LENGTH_KILOMETERS,
+    LENGTH_METERS,
+    LENGTH_CENTIMETERS,
+]
+
+MASS_UNITS = [
+    MASS_POUNDS,
+    MASS_OUNCES,
+    MASS_KILOGRAMS,
+    MASS_GRAMS,
+]
+
+VOLUME_UNITS = [
+    VOLUME_GALLONS,
+    VOLUME_FLUID_OUNCE,
+    VOLUME_LITERS,
+    VOLUME_MILLILITERS,
+]
+
+TEMPERATURE_UNITS = [
+    TEMP_FAHRENHEIT,
+    TEMP_CELSIUS,
+]
+
+NOT_RECOGNIZED_TEMPLATE = '{} is not a recognized {} unit.'  # type: str
+
+
+def is_valid_unit(unit: str, unit_type: str) -> bool:
+    """Check if the unit is valid for it's type."""
+    if unit_type == TYPE_LENGTH:
+        units = LENGTH_UNITS
+    elif unit_type == TYPE_TEMPERATURE:
+        units = TEMPERATURE_UNITS
+    elif unit_type == TYPE_MASS:
+        units = MASS_UNITS
+    elif unit_type == TYPE_VOLUME:
+        units = VOLUME_UNITS
+    else:
+        return False
+
+    return unit in units
+
+
+class UnitSystem(dict):
```

Comment:
We should never extend a base type. That's bad practice as now we have to support anything that `dict` supports.

Instead, let's make it a simple class. That way you can use attributes instead of requiring people to import a constant to get a unit. It also allows us to move conversions into the class.

```python
class UnitSystem(object):
    def __init__(self: dict, name: str, temperature: str, length: str,
                 volume: str, mass: str) -> None:
        """Initialize the unit system object."""
        self.temperature_unit = temperature

    def temperature(self, value, unit):
        """Convert temperature to user preferred unit."""
        [ … Move this function from Config to here … ]
```

Now we can access the temperature unit without an import:

```python
# Also rename unit_system to units, is shorter and makes more sense if people can customize some units
print(hass.config.unit.temperature_unit)
```